### ISSUE TITLE
Returning {} from generic async function doesn't produce error when it should

### ISSUE THREAD

**Issue by tcrognon** [Microsoft/TypeScript#14048]
**TypeScript Version:**  2.1.6
```ts
function foo<T>(bar: () => Promise<T>) {}

foo<{ required: string }>(async function () {
    return 'test'; // this is an error, as it should be
});

foo<string>(async function () {
    return { required: 'test' }; // this is an error, as it should be
});

foo<{ required: string }>(async function () {
    return {}; // this should be an error, but it isn't
});

foo<string>(async function () {
    return {}; // this should be an error, but it isn't
});

```

**Comment by mhegazy** [comment]
should be fixed in latest.
